algorithms/nautilus/fuzzer/fuzzer.cpp

```diff
@@ -0,0 +1,278 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+/**
+ * @file fuzzer.cpp
+ * @brief Fuzzing loop of Nautilus.
+ * @author Ricerca Security <fuzzuf-dev@ricsec.co.jp>
+ */
+#include <fstream>
+#include <memory>
+#include <nlohmann/json.hpp>
+#include <string>
+#include <vector>
+#include "fuzzuf/algorithms/nautilus/fuzzer/fuzzer.hpp"
+#include "fuzzuf/algorithms/nautilus/fuzzer/mutation_hierarflow_routines.hpp"
+#include "fuzzuf/algorithms/nautilus/fuzzer/other_hierarflow_routines.hpp"
+#include "fuzzuf/algorithms/nautilus/fuzzer/update_hierarflow_routines.hpp"
+#include "fuzzuf/hierarflow/hierarflow_intermediates.hpp"
+#include "fuzzuf/hierarflow/hierarflow_node.hpp"
+#include "fuzzuf/hierarflow/hierarflow_routine.hpp"
+#include "fuzzuf/logger/logger.hpp"
+#include "fuzzuf/utils/common.hpp"
+#include "fuzzuf/utils/filesystem.hpp"
+
+
+namespace fuzzuf::algorithm::nautilus::fuzzer {
+
+using json = nlohmann::json;
+
+/**
+ * @fn
+ * @brief Construct Nautilus fuzzer
+ * @param (state_ref) Reference to the state of Nautilus fuzzer
+ */
+NautilusFuzzer::NautilusFuzzer(std::unique_ptr<NautilusState>&& state_ref)
+  : state(std::move(state_ref)) {
+  /* Check files and load grammar */
+  CheckPathExistence();
+  LoadGrammar();
+
+  /* Construct fuzzing loop */
+  BuildFuzzFlow();
+
+  /* Clear screen */
+  MSG(TERM_CLEAR);
+}
+
+/**
+ * @fn
+ * @brief Build HierarFlow of Nautilus
+ */
+void NautilusFuzzer::BuildFuzzFlow() {
+  using fuzzuf::hierarflow::CreateNode;
+  using namespace fuzzuf::algorithm::nautilus::fuzzer::routine::other;
+  using namespace fuzzuf::algorithm::nautilus::fuzzer::routine::mutation;
+  using namespace fuzzuf::algorithm::nautilus::fuzzer::routine::update;
+
+  fuzz_loop = CreateNode<FuzzLoop>(*state);
+
+  /* Main flow */
+  auto select_input            = CreateNode<SelectInput>(*state);
+  auto process_chosen_input_or = CreateNode<ProcessInput>(*state);
+  auto generate_input          = CreateNode<GenerateInput>(*state);
+  auto update_state            = CreateNode<UpdateState>(*state);
+
+  /* Processing flow */
+  auto initialize_state_or = CreateNode<InitializeState>(*state);
+  auto apply_det_muts_or   = CreateNode<ApplyDetMuts>(*state);
+  auto apply_rand_muts     = CreateNode<ApplyRandMuts>(*state);
+
+  /* Mutation flow */
+  auto splice    = CreateNode<MutSplice>(*state);
+  auto havoc     = CreateNode<MutHavoc>(*state);
+  auto havoc_rec = CreateNode<MutHavocRec>(*state);
+
+  fuzz_loop << (
+    select_input << (
+      process_chosen_input_or
+      || generate_input // TODO: maybe execute.HardLink() here
```

Comment:
Probably, this kind of routines named `*_or` can be an irregular node and it would be more comprehensible.

Here, what we want is something like switch cases and currently you realize it with std::variant and "linear fallback"(if the state is an instance of `DetState,` then ~, else if the state ~, ...).
Instead of checking the current state linearly and across nodes, you can create one node that represents "switch jump table" and enumerate all the switch cases as its successors.
The example of HierarFlowIrregularNode is here:
https://github.com/fuzzuf/fuzzuf/blob/master/include/fuzzuf/hierarflow/utility.hpp#L119-L156
https://github.com/fuzzuf/fuzzuf/blob/master/test/hierarflow/utility.cpp#L70-L106

Reply:
Modified in [7d0f276](https://github.com/fuzzuf/fuzzuf/pull/23/commits/7d0f276234a9ee659dcd3fba4be32a3d990c4c7a)

Reply:
wonderful